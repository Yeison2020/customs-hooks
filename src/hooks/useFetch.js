import { useState, useEffect } from "react";
// Custom hook
const useFetch = (url) => {
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  useEffect(() => {
    // This part here Make the loading part to be shown on the screen It likes reseting to make it long
    setState({ data: null, loading: true, error: null });
    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        setState({
          data,
          loading: false,
          error: null,
        });
      });
  }, [url]);

  return state;
};

export default useFetch;
