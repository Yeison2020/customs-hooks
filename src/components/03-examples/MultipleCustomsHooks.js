import React from "react";
import "../02-useEffect/effects.css";
import useFetch from "../../hooks/useFetch";
import useCounter from "../../hooks/useCounter";

const MultipleCustomsHooks = () => {
  const { counter, increment } = useCounter(1);
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  // Why !!data && data[0]
  // data = null first time meaning
  // !data = true second time if It's
  // !!data = false If It's false meaning the data is not null display data at index data[0]

  const { author, quote } = !!data && data[0];
  console.log(author, quote);
  return (
    <div>
      <h1>BreakingBad Quotes</h1>
      <hr />

      {loading ? (
        <div className="alert alert-info text-center">Loading....</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p>{quote}</p>
          <p className="blockquote-footer">{author}</p>
        </blockquote>
      )}
      <button onClick={increment} className="btn btn-primary">
        Next quote
      </button>
    </div>
  );
};

export default MultipleCustomsHooks;
