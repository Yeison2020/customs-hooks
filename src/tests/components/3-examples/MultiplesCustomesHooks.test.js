import React from "react";
import { shallow } from "enzyme";
import Enzyme from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import MultipleCustomsHooks from "../../../components/03-examples/MultipleCustomsHooks";
import useFetch from "../../../hooks/useFetch";
// This libray simulate the result of differents customs Hooks.
jest.mock("../../../hooks/useFetch");
Enzyme.configure({ adapter: new Adapter() });

describe("Puebas en MultipleCostumsHooks", () => {
  test("Should show the correct result", () => {
    useFetch.mockReturnValue({
      data: null,
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomsHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Should show the information of the quote It", () => {
    useFetch.mockReturnValue({
      data: [{ author: "Yeison", quotes: "Hola mundo" }],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomsHooks />);
    expect(wrapper.find(".alert").exists()).toBe(false);
    console.log(wrapper.html());
  });
});
