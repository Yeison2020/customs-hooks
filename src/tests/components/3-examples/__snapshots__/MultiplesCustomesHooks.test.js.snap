// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Puebas en MultipleCostumsHooks Should show the correct result 1`] = `ShallowWrapper {}`;
